compliance/collection/nodeinventorizer/caching_inventorizer.go

```diff
@@ -0,0 +1,342 @@
+package nodeinventorizer
+
+import (
+	"encoding/json"
+	"os"
+	"time"
+
+	timestamp "github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/compliance/collection/metrics"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/jsonutil"
+	"github.com/stackrox/rox/pkg/uuid"
+	"github.com/stackrox/scanner/database"
+	scannerV1 "github.com/stackrox/scanner/generated/scanner/api/v1"
+	"github.com/stackrox/scanner/pkg/analyzer/nodes"
+	"golang.org/x/exp/maps"
+)
+
+// CachingScannerOpts control behaviour of all CachingScanner related functions
+type CachingScannerOpts struct {
+	InventoryCachePath  string              // Path to which a cached inventory is written to
+	BackoffFilePath     string              // Path to which the backoff file is written to
+	BackoffWaitCallback func(time.Duration) // Callback that gets called if a backoff file is found
+}
+
+// CachingScanner is an implementation of NodeInventorizer that keeps a local cache of results.
+//
+// To reduce strain on the Node, a linear backoff is checked before collecting an inventory.
+// Additionally, a cached inventory from an earlier invocation may be used instead of a full inventory run if it is fresh enough.
+// Note: This does not prevent strain in case of repeated pod recreation, as both mechanisms are based on an EmptyDir.
+type CachingScanner struct {
+	opts *CachingScannerOpts
+}
+
+// NewCachingScanner returns a ready to use instance of Caching Scanner
+func NewCachingScanner(inventoryCachePath string, backoffFilePath string, backoffCallback func(time.Duration)) *CachingScanner {
+	return &CachingScanner{opts: &CachingScannerOpts{
+		InventoryCachePath:  inventoryCachePath,
+		BackoffFilePath:     backoffFilePath,
+		BackoffWaitCallback: backoffCallback,
+	}}
+}
+
+// Scan scans the current node and returns the results as storage.NodeInventory struct
+func (c *CachingScanner) Scan(nodeName string) (*storage.NodeInventory, error) {
+	// check whether a cached inventory exists that is recent enough to use
+	cachedInventory, creationTime := readInventory(c.opts.InventoryCachePath)
+	if cachedInventory != nil && isCachedInventoryValid(creationTime) {
+		log.Debugf("Using cached node scan created at %v", creationTime)
+		return cachedInventory, nil
+	}
+
+	// check for existing backoff, wait for specified duration if needed, then persist the new backoff duration
+	initialBackoff := env.NodeScanInitialBackoff.DurationSetting()
```

Comment:
I think this value will not be changed after the start, so maybe we could move it into a field of the `CachingScanner` and call the `env.NodeScanInitialBackoff.DurationSetting()` only once in the constructor?

This would also be a step to making it easier to test.

This comment applies to all  `env.*.DurationSetting()` calls - consider to add all such vars as dependencies of this type and inject them over `func NewCachingScanner()`

Reply:
I agree, that this would simplify things quite a bit. 
On the other hand, we would introduce the need to restart the container if any of the env var are changed.
That might not so bad, as from experience this is quite a common thing.